FreeRTOS_ARP.c

```diff
@@ -305,9 +308,42 @@ static void vARPProcessPacketReply( ARPPacket_t * pxARPFrame,
                                     uint32_t ulSenderProtocolAddress )
 {
     ARPHeader_t * pxARPHeader = &( pxARPFrame->xARPHeader );
+    static BaseType_t uxCount = 0;
 
     iptracePROCESSING_RECEIVED_ARP_REPLY( ulTargetProtocolAddress );
     vARPRefreshCacheEntry( &( pxARPHeader->xSenderHardwareAddress ), ulSenderProtocolAddress, pxTargetEndPoint );
+
+    if( pxARPWaitingNetworkBuffer != NULL )
+    {
+        IPPacket_t * pxARPWaitingIPPacket = ipCAST_PTR_TO_TYPE_PTR( IPPacket_t, pxARPWaitingNetworkBuffer->pucEthernetBuffer );
+        IPHeader_t * pxIPHeader = &( pxARPWaitingIPPacket->xIPHeader );
+
+        if( ulSenderProtocolAddress == pxIPHeader->ulSourceIPAddress )
+        {
+            IPStackEvent_t xEventMessage;
+            const TickType_t xDontBlock = ( TickType_t ) 0;
+
+            xEventMessage.eEventType = eNetworkRxEvent;
+            xEventMessage.pvData = ( void * ) pxARPWaitingNetworkBuffer;
+
+            xSendEventStructToIPTask( &xEventMessage, xDontBlock );
```

Comment:
That is a great solution! The IP-task will keep the network buffer until a matching ARP request comes in. At that moment, the message will be processed.

I tested it again, and the PCAP looks like expected. Here is a PCAP: [icmp_test.zip](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/files/7088483/icmp_test.zip). The delayed reply comes immediately after the ARP messages.

This is very simple:
~~~c
    uxCount++;
    if( uxCount > 5 )
    {
        /* Release the network buffer. */
    }
~~~
Maybe an alarm clock would be nicer? There is a simple light-weight timer `IPTimer_t`, like e.g. :
~~~c
    /** @brief ARP timer, to check its table entries. */
    static IPTimer_t xARPTimer;

    /* Reload the timer. */
    prvIPTimerReload( &xARPTimer, pdMS_TO_TICKS( ipARP_TIMER_PERIOD_MS ) );

   /* Check if it expires. */
    if( prvIPTimerCheck( &xARPTimer ) != pdFALSE )
    {
        ( void ) xSendEventToIPTask( eARPTimerEvent );
    }

    /* See how long the IP-task may sleep. */
    if( xARPTimer.bActive != pdFALSE_UNSIGNED )
    {
        if( xARPTimer.ulRemainingTime < xMaximumSleepTime )
        {
            xMaximumSleepTime = xARPTimer.ulRemainingTime;
        }
    }
~~~

Reply:
Yes, I was thinking along the same lines. I was going to add this along with a timer or create a new one which would be checked by the IP task.
I shall add more commits soon (tomorrow)

Reply:
I have added more commits and added UDP processing as well. Thank you.